src/main/java/uk/gov/caz/psr/service/GetAndUpdatePaymentsService.java

```
@@ -31,21 +33,28 @@
   public Optional<Payment> getExternalPaymentAndUpdateStatus(UUID id) {
     Preconditions.checkNotNull(id, "ID cannot be null");
 
-    Payment internalPayment = internalPaymentsRepository.findById(id).orElse(null);
+    Payment payment = internalPaymentsRepository.findById(id).orElse(null);
```

Sorry for perhaps silly question: how is it possible that we get ID and do not find any payment in database? Is this valid use case from user perspective? My understanding is that the only way for this to happen is when frontend passes invalid UUID, right? Should we propagate this 'error' by returning Optional? I would expect to fail fast, throw exception here and return 400 to the caller.... Or am I missing something (most probably)?

---

It's not a silly question, you're right about it and not missing anything: from the user perspective it is not a valid case as this endpoint completes the payment journey, hence the payment has to be present in the database. The whole issue comes down to error propagation, whether we want to throw exception(s) or return some value. Initially I chose the latter, maybe because of the automatic inclinations that when a method starts with `get`, then `Optional` is a good candidate as a returned value. Currently controller checks if Optional is empty and returns `404` if it is, `200` code otherwise.

It's slightly a matter of taste, in my opinion. If you think the urgency of the error will be better reflected to the reader (I guess on the FE side they don't really care if we return `4xx` or `5xx` status code) by throwing an exception, I will be happy to change it.

---

Thanks for very detailed explanation. It makes sense. I still can't get my head around how to properly handle exceptions in Java (I guess I will have to live my life with this uncertainty). Your solution is very good, and I'm OK with repositories being dump wrappers around DB - they don't see a row? They return Optional.empty. Simple. Caller is still obliged to handle it in a proper manner - maybe throw Exception, maybe skip etc. So ok for me!